### ISSUE TITLE
Add a way to predict or estimate expected tokens or expected costs

### ISSUE THREAD

**Issue by davidberenstein1957** [argilla-io/distilabel#462]
**Is your feature request related to a problem? Please describe.**
Generation with reliance on external APIs might be expensive.

**Describe the solution you'd like**
Something like a pipeline method `expected_costs` or `expected_tokens`

**Describe alternatives you've considered**
Wait and see.
**Additional context**
Perhaps something like `expected_time` would also be interesting?

**Comment by strickvl** [comment]
All these would be really useful, I think. You'd want:

- expected costs for a particular pipeline (either calculated directly, or via directing to some API or online costs calculator for the particular LLMs involved). In the case of multiple LLMs involved, this might get complicated i.e.
- expected tokens to be used for a particular pipeline, and split by LLM / task (so you can manually make the calculation yourself if you want
- add a 'max_budget' option for a param, which would either stop when you hit that estimated cost, or it wouldn't run the param and would suggest ways to downscale the amount of work your pipeline does in order to stay within the costs?
- expected_time is maybe interesting, but probably hard to do. But I agree it'd be pretty useful if you could get a ballpark figure.

You might even consider abstracting this out into some online calculator tool where you just get the number of tokens for a raw data file and then for a typical task you could get a rough estimate of both time and cost for typical scenarios (i.e. gpt3.5 or claude-haiku etc).

I think this'd all be really beneficial for the tool. In the end, you can only make very loose guesses right now based on small subsection runs and then try to extrapolate out.

**Comment by DiTo97** [comment]
@strickvl, @davidberenstein1957, any progress on cost tracking?

In both personal and enterprise work, I have switched to Distilabel for synthetic data generation, mostly using OpenAI models, but one thing I really miss with respect to the custom SDG pipeline I had before is cost tracking.

This is even more important at the enterprise level to justify the cost of generating datasets within our model training workflows. SDG is used a lot as we often work with private data that we can rarely train on. I know how tricky it is to manage both open and private models with the frequent changes in pricing, but just wanted to know if you had given any thought to it.

As I said, I mostly use OpenAI, and seldom HF Inference Endpoints. If you haven't made any progress due to the known complexity, is there a way to intercept the completion usage of generations so that we could still do it ourselves?

In the past I have used something as simple as:

```python
import typing_extensions

from openai.types import CompletionUsage


class Pricing(typing_extensions.TypedDict):
    cached: float
    prompt: float
    completion: float
    conversion: int
    currency: str


class Tracing(typing_extensions.TypedDict):
    cached_tokens: int
    prompt_tokens: int
    completion_tokens: int
    costing: float
    currency: str


_pricing_LUT = {
    "o3-mini": dict(
        cached=0.55,
        prompt=1.10,
        completion=4.40,
        conversion=1_000_000,
        currency="$"
    )
}


def compute_usage_and_costing(metaset: list[CompletionUsage], model: str) -> Tracing:
    if model not in _pricing_LUT:
        content = f"The pricing for model {model} is not available."
        raise ValueError(content)

    pricing = _pricing_LUT[model]

    cached_tokens = 0
    prompt_tokens = 0
    completion_tokens = 0

    for usage in metaset:
        cached_tokens += usage.prompt_tokens_details.cached_tokens
        prompt_tokens += usage.prompt_tokens
        completion_tokens += usage.completion_tokens

    return {
        "cached_tokens": cached_tokens,
        "prompt_tokens": prompt_tokens,
        "completion_tokens": completion_tokens,
        "costing": sum((
            pricing["prompt"] * (prompt_tokens - cached_tokens),
            pricing["cached"] * cached_tokens,
            pricing["completion"] * completion_tokens 
        )) / pricing["conversion"],
        "currency": pricing["currency"]
    }
```

For the time being, I have been creating a brand new OpenAI API key for each new pipeline, so I still have a way to measure the cost as I know that each key is only tied to a specific data generation pipeline, but it's not very convenient.

**Comment by davidberenstein1957** [comment]
I agree this would be a great feature. Perhaps @plaguss and @gabrielmbmb can take a look. I am not sure if it has been scheduled already?
